### ISSUE TITLE
Type error in one file, only after opening other.

### ISSUE THREAD

**Issue by lsmor** [haskell/haskell-language-server#3835]
### Your environment

- OS: Ubuntu
- ghc-9.2.8 from ghcup
- The project is current master branch of `ghcup` tool. (commit: a73ce18)
- vscode; haskell extension
- bug reproduced with `hls-2.1.0.0` and `2.3.0.0`

```yaml
# hie.yaml
cradle:
  cabal:
  - component: "ghcup:lib:ghcup"
    path: ./lib
  - component: "ghcup:exe:ghcup"
    path: ./app/ghcup
  - component: "ghcup:test:ghcup-test"
    path: ./test/ghcup-test
  - component: "ghcup:test:ghcup-optparse-test"
    path: ./test/optparse-test
```

### Steps to reproduce

1. Clone `ghcup` repo (commit at the time of writting this: a73ce18)
2. Open `app/BrickMain.hs` file  in the editor (vscode in my case)
3. Wait until `hls` finishes compiling/indexing etc... and stays idle
4. Without closing `BrickMain.hs`  open `lib/Types.hs` (the file not the folder)

### Expected behaviour

Nothing should happen other than listing / processing / indexing files.

### Actual behaviour

Suddenly,  file `BrickMain.hs` shows a type error in line 577. Very rare since the project does compile perfectly with `cabal`

**Comment by fendor** [comment]
Hi!

Without checking too thoroughly, my gut tells me this is related to #3738. It is usually the culprit for such things.

**Comment by fendor** [comment]
I can reproduce and the issue is caused by #3738. Essentially, you have to make sure maintain the closure property of home units, see https://gitlab.haskell.org/ghc/ghc/-/blob/master/compiler/GHC/Unit/Env.hs#L548.

Make you sure you open all components in the ghcup project in your project to fix this or wait for #3738. Closing this in favour of #3738 which tracks our efforts to fix these kinds of issues.
